pennylane/templates/subroutines/hilbert_schmidt.py

```diff
@@ -0,0 +1,256 @@
+
+            @qml.qnode(dev)
+            def hilbert_test(v_params, v_function, v_wires, u_tape):
+                qml.HilbertSchmidt(v_params, v_function=v_function, v_wires=v_wires, u_tape=u_tape)
+                return qml.probs(u_tape.wires + v_wires)
+
+            def cost_hst(parameters, v_function, v_wires, u_tape):
+                return (1 - hilbert_test(v_params=parameters, v_function=v_function, v_wires=v_wires, u_tape=u_tape)[0])
+
+        Now that you have defined your cost function you can call it for specific parameters.
+
+        >>> cost_hst([0], v_function = v_circuit, v_wires = [1], u_tape = U)
+        1
+
+
+    **Reference**
+
+    [1] Sumeet Khatri, Ryan LaRose, Alexander Poremba, Lukasz Cincio, Andrew T. Sornborger and Patrick J. Coles
+    Quantum-assisted Quantum Compiling.
+    `arxiv/1807.00800 <https://arxiv.org/pdf/1807.00800.pdf>`_
+    """
+    num_wires = AnyWires
+    grad_method = None
+
+    def __init__(self, *params, v_function, v_wires, u_tape, do_queue=True, id=None):
+
+        if not isinstance(u_tape, qml.tape.QuantumTape):
+            raise qml.QuantumFunctionError("The argument u_tape must be a QuantumTape.")
+
+        u_wires = u_tape.wires
+        self.hyperparameters["u_tape"] = u_tape
+
+        if not callable(v_function):
+            raise qml.QuantumFunctionError(
+                "The argument v_function must be a callable quantum function."
+            )
+
+        self.hyperparameters["v_function"] = v_function
+
+        v_tape = qml.transforms.make_tape(v_function)(*params)
+        self.hyperparameters["v_tape"] = v_tape
+        self.hyperparameters["v_wires"] = v_tape.wires
+
+        if len(u_wires) != len(v_wires):
+            raise qml.QuantumFunctionError("U and V must have the same number of wires.")
+
+        if not qml.wires.Wires(v_wires).contains_wires(v_tape.wires):
+            raise qml.QuantumFunctionError("All wires in v_tape must be in v_wires.")
+
+        # Intersection of wires
+        if len(qml.wires.Wires.shared_wires([u_tape.wires, v_tape.wires])) != 0:
+            raise qml.QuantumFunctionError("u_tape and v_tape must act on distinct wires.")
+
+        wires = qml.wires.Wires(u_wires + v_wires)
+
+        super().__init__(*params, wires=wires, do_queue=do_queue, id=id)
+
+    @property
+    def num_params(self):
+        return 1
+
+    @staticmethod
+    def compute_decomposition(
+        params, wires, u_tape, v_tape, v_function=None, v_wires=None
+    ):  # pylint: disable=arguments-differ,unused-argument
+        r"""Representation of the operator as a product of other operators (static method)."""
+        n_wires = len(u_tape.wires + v_tape.wires)
+        decomp_ops = []
+
+        first_range = range(0, int(n_wires / 2))
+        second_range = range(int(n_wires / 2), n_wires)
+
+        # Hadamard first layer
+        for i, wire in enumerate(wires):
+            if i in first_range:
```

Comment:
How about iterating over `first_range` and indexing into `wires` via `wires[i]`? That should save the `if` statement and having to iterate over all the wires.

Reply:
Yes I will change it!